akkoma: init at 3.5.0 (#192285)
NixOS/nixpkgs: nixos/modules/services/web-apps/akkoma.md

store by setting the configuration option to an attribute set containing the attribute
{option}`_secret` – a string pointing to the file containing the actual value of the option.

For the mandatory configuration settings these secrets will be generated automatically if the
referenced file does not exist during startup, unless disabled through
{option}`services.akkoma.initSecrets`.

The following configuration binds Akkoma to the local IPv6 address `::1` on port 4000, expecting
to be run behind a HTTP proxy on `fediverse.example.com`.

```nix
services.akkoma.enable = true;
services.akkoma.config = {
  ":pleroma" = {
    ":instance" = {
      name = "My Akkoma instance";
      description = "More detailed description";
      email = "admin@example.com";
      registration_open = false;
    };

    "Pleroma.Web.Endpoint" = {
      url.host = "fediverse.example.com";
      http = {
        ip = "::1";
        port = 4000;
      };
    };
  };
};
```

Please refer to the [configuration cheat sheet](https://docs.akkoma.dev/stable/configuration/cheatsheet/)
for additional configuration options.

> The service can be enabled with the usual
> 
> ```ShellSession
> $ nixos-rebuild switch
> ```

Reviewer: ```suggestion
```

The service is enabled by ``services.akkoma.enable = true;`` and we do not need to explain and assume how the nixos system is build here.

## User management {#modules-services-akkoma-user-management}

After the Akkoma service is running, the administration utility can be used to
[manage users](https://docs.akkoma.dev/stable/administration/CLI_tasks/user/). In particular an
administrative user can be created with

```ShellSession

> $ sudo -u akkoma RELEASE_COOKIE=/run/akkoma/COOKIE \
> pleroma_ctl user new <nickname> <email> --admin --moderator --password <password>

Reviewer: ```suggestion
$ sudo -u akkoma RELEASE_COOKIE=/run/akkoma/COOKIE pleroma_ctl user new <nickname> <email> --admin --moderator --password <password>
```

```

## Proxy configuration {#modules-services-akkoma-proxy-configuration}

> Although it is possible to expose Akkoma directly, it is common practice to operate it behind an
> HTTP reverse proxy such as HAProxy or nginx.

Reviewer: ```suggestion
Although it is possible to expose Akkoma directly, it is common practice to operate it behind an
HTTP reverse proxy such as nginx.
```
We do not have good support haproxy.

Author: I noticed. Maybe that’s something I may have a look at once I am through with this.

```nix

> security.acme = {
>   email = "admin@example.com";
>   acceptTerms = true;
> };
> 
> services.akkoma.nginx = {
>   enableACME = true;
>   forceSSL = true;
> };
> 
> services.nginx = {
>   enable = true;
> 
>   clientMaxBodySize = "16m";
>   recommendedTlsSettings = true;
>   recommendedOptimisation = true;
>   recommendedGzipSettings = true;
> };

Reviewer: This is incomplete and at least missing a domain for the cert. Instead of having an incomplete example I'd much prefer to redirect people to setup ACME.

```suggestion
services.akkoma.nginx = {
  enableACME = true;
  forceSSL = true;
};

services.nginx = {
  enable = true;

  clientMaxBodySize = "16m";
  recommendedTlsSettings = true;
  recommendedOptimisation = true;
  recommendedGzipSettings = true;
};
```

```

## Frontend management {#modules-services-akkoma-frontend-management}

Akkoma will be deployed with the `pleroma-fe` and `admin-fe` frontends by default. These can be
overridden by setting the {option}`services.akkoma.frontends` option.

The following example overrides the primary frontend’s default configuration using a custom
derivation.

```nix
services.akkoma.frontends.primary.package = let
  default-config = pkgs.writeText "default-config.json" (builtins.toJSON {
    expertLevel = 1;
    collapseMessageWithSubject = false;
    stopGifs = false;
    replyVisibility = "following";
    webPushHideIfCW = true;
    hideScopeNotice = true;
    renderMisskeyMarkdown = false;
    hideSiteFavicon = true;
    postContentType = "text/markdown";
    showNavShortcuts = false;

>   });;

Reviewer: Is there any advantage to first translating the json to nix instead of using json direct? Also why do we write this to the store first? We could directly inject that into jq.

Author: I copied this from another user without giving it too much thought. But I agree that writing the JSON to the store first is unnecessary – I suspect that the original author was unfamiliar with `passAsFile`. I personally feel that writing an attribute set is somewhat more comfortable than a string of JSON.

in pkgs.runCommand "pleroma-fe" {
  nativeBuildInputs = with pkgs; [ jq ];
} ''

>   cp -R --reflink=auto ${pkgs.akkoma-frontends.pleroma-fe} $out

Reviewer: Why not use nested symlinks we can modify with lndir?

Author: Since the service uses bind mounts to merge the frontends into the state directory, that would work as long as the symlink targets are absolute paths and confinement is disabled.

With confinement enabled, the symlinks would likely point to paths which don’t exist in the mount namespace, unless added explicitly.

Would hard links be a suitable alternative?

```Shell
cp -lR ${pkgs.akkoma-frontends.pleroma-fe} $out
rm $out/static/config.json
jq -s add ${pkgs.akkoma-frontends.pleroma-fe}/static/config.json ${config} >$out/static/config.json
```

Reviewer: I don't think that would make much of a difference.

  jq -s add ${pkgs.akkoma-frontends.pleroma-fe}/static/config.json ${default-config} \
    >$out/static/config.json
'';
```

## Advanced deployment options {#modules-services-akkoma-advanced-deployment}

### Confinement {#modules-services-akkoma-confinement}

> The Akkoma systemd service may be confined to a chroot with

Reviewer: We should add a warning here and what implications this has and why it is not enabled by default.

```nix
services.systemd.akkoma.confinement.enable = true;
```

### Distributed deployment {#modules-services-akkoma-distributed-deployment}

> Being an Elixir application, Akkoma can be deployed in a
> [distributed](https://www.erlang.org/doc/reference_manual/distributed.html) fashion.
> 
> This requires setting {option}`services.akkoma.dist.address` and
> {option}`services.akkoma.dist.cookie`. The specifics depend strongly on the deployment environment
> and are outside of the scope of this configuration guide.

Reviewer: ```suggestion
Being an Elixir application, Akkoma can be deployed in a distributed fashion.

This requires setting {option}`services.akkoma.dist.address` and
{option}`services.akkoma.dist.cookie`. The specifics strongly depend on the deployment environment. For more information please visit the [erlang documentation](https://www.erlang.org/doc/reference_manual/distributed.html) about it. 
```

For the mandatory configuration settings these secrets will be generated automatically if the
referenced file does not exist during startup, unless disabled through
[{option}`services.akkoma.initSecrets`](options.html#opt-services.akkoma.initSecrets).

The following configuration binds Akkoma to the Unix socket `/run/akkoma/socket`, expecting to
be run behind a HTTP proxy on `fediverse.example.com`.

```nix
services.akkoma.enable = true;
services.akkoma.config = {
  ":pleroma" = {
    ":instance" = {
      name = "My Akkoma instance";
      description = "More detailed description";
      email = "admin@example.com";
      registration_open = false;
    };

    "Pleroma.Web.Endpoint" = {
      url.host = "fediverse.example.com";
    };
  };
};
```

Please refer to the [configuration cheat sheet](https://docs.akkoma.dev/stable/configuration/cheatsheet/)
for additional configuration options.

## User management {#modules-services-akkoma-user-management}

After the Akkoma service is running, the administration utility can be used to
[manage users](https://docs.akkoma.dev/stable/administration/CLI_tasks/user/). In particular an
administrative user can be created with

```ShellSession

> $ sudo -u akkoma pleroma_ctl user new <nickname> <email> --admin --moderator --password <password>

Reviewer: This could be made more convenient by adding a wrapper like nextcloud does.

Author: Done as requested.

### Media proxy {#modules-services-akkoma-media-proxy}

Without the media proxy function, Akkoma does not store any remote media like pictures or video
locally, and clients have to fetch them directly from the source server.

```nix
# Enable media proxy
services.akkoma.config.":pleroma".":media_proxy" = {
  enabled = true;
  base_url = "https://cache.fediverse.example.com";
  proxy_opts.redirect_on_failure = true;
};

# Adjust the persistent cache size as needed:
#  Assuming an average object size of 128 KiB, around 1 MiB
#  of memory is required for the key zone per GiB of cache.
# Ensure that the cache directory exists and is writable by nginx.
services.nginx.commonHttpConfig = ''
  proxy_cache_path /var/cache/nginx/cache/akkoma-media-cache
    levels= keys_zone=akkoma_media_cache:16m max_size=16g
    inactive=1y use_temp_path=off;
'';

nginx.virtualHosts."cache.fediverse.example.com" = {
  locations."/" = {
    proxyPass = "http://unix:/run/akkoma/socket";

    extraConfig = ''
      proxy_cache akkoma_media_cache;

      # Cache objects in slices of 1 MiB
      slice 1m;
      proxy_cache_key $host$uri$is_args$args$slice_range;
      proxy_set_header Range $slice_range;

>       # Decouple client and upstream requests

Reviewer 2: ```suggestion
      # Decouple client and upstream responses
```

Author: Corrected to: *Decouple proxy and upstream responses*

>       proxy_buffering on;

Reviewer: This is recommended against by nginx itself because slow clients can DOS the server.

Author: Do you have more information on that?

According to the module reference it is [on by default](https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_buffering) and this [blog post](https://www.nginx.com/blog/avoiding-top-10-nginx-configuration-mistakes/#proxy_buffering-off) suggests that disabling it might disable caching altogether, which would defeat the purpose of running a cache. 

The comment should probably state that the client and upstream *responses* are decoupled.

Reviewer 2: I made a suggestion here: https://github.com/NixOS/nixpkgs/pull/192285/files#r1014412099
